### ISSUE TITLE
Push/Pop FS:00 handlers are not thread safe

### ISSUE THREAD

**Issue by JohnGodgames** [Cxbx-Reloaded/Cxbx-Reloaded#986]
Tested by https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/93cf2ab789aaa422ef8f3a194b45f22e6f990731

## Issued games:
**Avatar: The Last Airbender
Genma Onimusha
Guilty Gear Isuka
Muzzle Flash
The Incredible Hulk: Ultimate Destruction**

https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/blob/master/src/CxbxKrnl/EmuFS.cpp#L379

```
__declspec(naked) void EmuFS_PushDwordPtrFs00()
{
	uint32 returnAddr;
	uint32 temp;

	__asm
	{
		call EmuFS_RefreshKPCR
		►►►pop returnAddr
		mov temp, eax
		mov eax, fs : [TIB_ArbitraryDataSlot]
		push dword ptr [eax]
		mov eax, temp
		push returnAddr
		ret
	}
}
```

**Comment by Nukem9** [comment]
It's because of stack variables being used, but the [prologue code is omitted in naked functions](https://msdn.microsoft.com/en-us/library/h5w10wxs.aspx).

```
.text:0413A990 ; void __cdecl EmuFS_PushDwordPtrFs00(void)
.text:0413A990
.text:0413A990 temp            = dword ptr -8
.text:0413A990 returnAddr      = dword ptr -4
.text:0413A990
.text:0413A990                 call    ?EmuFS_RefreshKPCR@@YAXXZ ; EmuFS_RefreshKPCR(void)
.text:0413A995                 pop     dword ptr [ebp-4]
.text:0413A998                 mov     [ebp-8], eax
.text:0413A99B                 mov     eax, large fs:14h
.text:0413A9A1                 push    dword ptr [eax]
.text:0413A9A3                 mov     eax, [ebp-8]
.text:0413A9A6                 push    dword ptr [ebp-4]
.text:0413A9A9                 retn
```

EmuFS_PopDwordPtrFs00 will also be broken I guess.

**Comment by PatrickvL** [comment]
Moving the variables down should fix this problem.

But there's another: writing values to memory could lead to a concurrency collision (it's not thread safe). I proposed an alternative implementation a while ago to Luke, but I'd have to look it up....

**Comment by PatrickvL** [comment]
call EmuFS_RefreshKPCR

push 0 // reserve a slot at [esp-4] (return address will be copied to here)
push eax // backup eax
mov eax, [esp - 8] // read current return addr
mov [esp - 4], eax // write it to the reserved slot
mov eax, fs : [TIB_ArbitraryDataSlot] // read FS
mov [esp - 8], eax // overwrite the return address
pop eax // restore eax - stack now points to the slot where we moved our return address
ret

**Comment by Nukem9** [comment]
Would something like this work (no eax needed)? I'm just writing it off the top of my head....it'd be nice if these used a direct jmp instead of call. It'd eliminate those extra read/writes to esp.

```
push [esp]    ; Ret addr is now on the stack twice
add esp, 8    ; Move to slot where original push fs would write
push fs:[TIB_ArbitraryDataSlot]
add esp, 4    ; Move to duplicated ret addr
retn
```

**Comment by PatrickvL** [comment]
@Nukem9 according to Luke, your code crashes outright, care to investigate?

**Comment by Nukem9** [comment]
I should've just deleted that - there is no way to avoid using a scratch register here. It requires a pointer dereference (fs:[14h]) and then another dereference which can't be done on the stack.

If I'm reading this correctly:
```
; Original code https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/986#issue-304619151
;
mov eax, fs : [TIB_ArbitraryDataSlot] ; 1st dereference
push dword ptr [eax]                  ; 2nd dereference
```

```
; Your code (broken?) https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/986#issuecomment-373204458
;
mov eax, fs : [TIB_ArbitraryDataSlot] ; 1st deref
mov [esp - 8], eax                    ; Missing a deref
```

```
; My code (broken) https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/986#issuecomment-373217243
;
push fs:[TIB_ArbitraryDataSlot] ; 1st deref
add esp, 4                      ; Missing a deref
```

**Comment by JohnGodgames** [comment]
It's fixedy by (This solution is NOT thread safe, but is better than outright
crashing.) https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/992/commits/985bb16515b7e2301f8d0d939bb4e74826c0e8da close or wait for better solution?

**Comment by PatrickvL** [comment]
Good to hear the crash is resolved, thanks! However, the current solution isn't thread safe, so we'd better keep this issue open to devise a better solution for fixing both push and pop.

**Comment by LukeUsher** [comment]
Updated the issue title to reflect current status

**Comment by LukeUsher** [comment]
I have a potential fix for this which involves a lightweight assembly spinlock implementation, carefully crafted so it doesn't trash any registers. Expect a PR in the coming days.

**Comment by LukeUsher** [comment]
Damn, forgot about this, I'll get it sorted

**Comment by ergo720** [comment]
Fixed in #1345 and #1355.
